### ISSUE TITLE
分类的精度很低The accuracy of classification is very low

### ISSUE THREAD

**Issue by liyhc** [hyzhang98/SGNN#1]
~$ python run_classification.py 

============ Settings ============
Totally 3 layers:
1-th layer: Neurons: 128, inner_activation: <linear>, activations: <function leaky_relu> with {'negative_slope': 0.2}  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 20, 'lam': 0.001, 'batch_size': 2708} type: EGCN
2-th layer: Neurons: 64, inner_activation: <linear>, activations: <function relu>  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 20, 'lam': 0.001, 'batch_size': 2708} type: EGCN
3-th layer: Neurons: 32, inner_activation: <linear>, activations: <linear>  with parameters: {'learning_rate': 0.01, 'order': 2, 'max_iter': 20, 'lam': 0.001, 'batch_size': 140} type: EGCN
overlook_rates=[0.0, 0.0, 0.0]
BP_count=5, eta=100

============ Start Training ============
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 3.64663
iteration: 19, loss: 1.83314
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 2.91152
iteration: 19, loss: 0.29831
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 3.41604
iteration: 19, loss: 0.20162

======= ACC: 0.8240, F1-Score: 0.8184 =======

 Start the 0-th backward training
---------------- Start training the 2-th GNN (BACKWARD)
iteration:  0, loss: 0.19912
iteration: 19, loss: 0.16590
---------------- Start training the 1-th GNN (BACKWARD)
iteration:  0, loss: 18957.21484
iteration: 19, loss: 7881.58496
---------------- Start training the 0-th GNN (BACKWARD)
iteration:  0, loss: 33587.48438
iteration: 19, loss: 5992.19336

 Start the 1-th forward training
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 5777.09961
iteration: 19, loss: 1571.77930
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 8729.75977
iteration: 19, loss: 6437.59473
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 0.19442
iteration: 19, loss: 0.13605

======= ACC: 0.8180, F1-Score: 0.8139 =======

 Start the 1-th backward training
---------------- Start training the 2-th GNN (BACKWARD)
iteration:  0, loss: 0.13596
iteration: 19, loss: 0.15102
---------------- Start training the 1-th GNN (BACKWARD)
iteration:  0, loss: 23448.57422
iteration: 19, loss: 11046.17578
---------------- Start training the 0-th GNN (BACKWARD)
iteration:  0, loss: 100111.97656
iteration: 19, loss: 16950.33789

 Start the 2-th forward training
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 15371.56543
iteration: 19, loss: 1925.00269
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 11462.17188
iteration: 19, loss: 10080.41602
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 0.14696
iteration: 19, loss: 0.12255

======= ACC: 0.8170, F1-Score: 0.8118 =======

 Start the 2-th backward training
---------------- Start training the 2-th GNN (BACKWARD)
iteration:  0, loss: 0.12054
iteration: 19, loss: 0.13364
---------------- Start training the 1-th GNN (BACKWARD)
iteration:  0, loss: 22288.29492
iteration: 19, loss: 24405.45312
---------------- Start training the 0-th GNN (BACKWARD)
iteration:  0, loss: 781753.25000
iteration: 19, loss: 370540.03125

 Start the 3-th forward training
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 358461.50000
iteration: 19, loss: 201013.42188
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 31995.22266
iteration: 19, loss: 27575.53516
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 1.00401
iteration: 19, loss: 0.25730

======= ACC: 0.8320, F1-Score: 0.8179 =======

 Start the 3-th backward training
---------------- Start training the 2-th GNN (BACKWARD)
iteration:  0, loss: 0.25101
iteration: 19, loss: 0.11845
---------------- Start training the 1-th GNN (BACKWARD)
iteration:  0, loss: 5956.25684
iteration: 19, loss: 17832.21680
---------------- Start training the 0-th GNN (BACKWARD)
iteration:  0, loss: 8177320.00000
iteration: 19, loss: 6484571.50000

 Start the 4-th forward training
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 6409283.00000
iteration: 19, loss: 5041302.00000
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 16664.57031
iteration: 19, loss: 16367.78125
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 2.05036
iteration: 19, loss: 2.01811

======= ACC: 0.1300, F1-Score: 0.0329 =======

 Start the 4-th backward training
---------------- Start training the 2-th GNN (BACKWARD)
iteration:  0, loss: 2.01753
iteration: 19, loss: 1.98913
---------------- Start training the 1-th GNN (BACKWARD)
iteration:  0, loss: 0.14733
iteration: 19, loss: 0.15775
---------------- Start training the 0-th GNN (BACKWARD)
iteration:  0, loss: 12307708.00000
iteration: 19, loss: 10412658.00000

 Start the 5-th forward training
---------------- Start training the 0-th GNN forward
iteration:  0, loss: 10325490.00000
iteration: 19, loss: 8679510.00000
---------------- Start training the 1-th GNN forward
iteration:  0, loss: 1111.91064
iteration: 19, loss: 0.14132
---------------- Start training the 2-th GNN forward
iteration:  0, loss: 2.01080
iteration: 19, loss: 2.01000

======= ACC: 0.1300, F1-Score: 0.0329 =======

============ End Training ============

============ Settings ============
Totally 3 layers:
1-th layer: Neurons: 128, inner_activation: <linear>, activations: <function leaky_relu> with {'negative_slope': 0.2}  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 20, 'lam': 0.001, 'batch_size': 2708} type: EGCN
2-th layer: Neurons: 64, inner_activation: <linear>, activations: <function relu>  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 20, 'lam': 0.001, 'batch_size': 2708} type: EGCN
3-th layer: Neurons: 32, inner_activation: <linear>, activations: <linear>  with parameters: {'learning_rate': 0.01, 'order': 2, 'max_iter': 20, 'lam': 0.001, 'batch_size': 140} type: EGCN
overlook_rates=[0.0, 0.0, 0.0]
BP_count=5, eta=100

============ Start testing ============

======= ACC: 0.1429, F1-Score: 0.0357 =======

======= ACC: 0.1220, F1-Score: 0.0311 =======

======= ACC: 0.1300, F1-Score: 0.0329 =======

我在core数据集运行run_classification.py ，但精度很低，达不到论文报告的精度，这是什么原因？
I am running run_classification. py on the core dataset, but the accuracy is very low and cannot reach the accuracy reported in the paper. What is the reason for this?

**Comment by hyzhang98** [comment]
感谢关注。

根据我的经验过度训练GNN比较容易遇到模型崩溃的问题，从log中也能看到，在前两次BT时 (也就是3次FT和2次BT)，ACC还是正常的；这取决于每个GNN被更新多少次，以及lr设置。可以看到在第4次FT时，最后一个GNN loss已经变大了，而前3次都是没问题的，符合文章Fig. 7，说明参数已经出了问题。

抱歉代码比较乱，也疏于整理；我们在cora上的setting不是代码中默认设置的 (增加实验导致代码各版本比较混乱)，我们找一下当时的setting然后更新一下。

**Comment by liyhc** [comment]
好的，你的这个工作对我来说确实很新颖，我接触图网络不久，准备在这个工作之上实现一些想法，期待更新

**Comment by aWeirdName** [comment]
You can try the following settings：

**layers = [
    LayerParam(128, inner_act=linear_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN,
               learning_rate=10**-2, order=1, max_iter=60, lam=10**-3, batch_size=2708),
    LayerParam(64, inner_act=linear_func, act=relu_func, gnn_type=LayerParam.EGCN,
               learning_rate=10**-2, order=1, max_iter=60, lam=10**-3, batch_size=2708),
    LayerParam(32, inner_act=linear_func, act=linear_func, gnn_type=LayerParam.EGCN,
               learning_rate=0.01, order=2, max_iter=60, lam=10**-3, batch_size=140),
]**

and this is the result:

============ Settings ============
Totally 3 layers:
1-th layer: Neurons: 128, inner_activation: <linear>, activations: <function leaky_relu> with {'negative_slope': 0.2}  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 60, 'lam': 0.001, 'batch_size': 2708} type: EGCN
2-th layer: Neurons: 64, inner_activation: <linear>, activations: <function relu>  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 60, 'lam': 0.001, 'batch_size': 2708} type: EGCN
3-th layer: Neurons: 32, inner_activation: <linear>, activations: <linear>  with parameters: {'learning_rate': 0.01, 'order': 2, 'max_iter': 60, 'lam': 0.001, 'batch_size': 140} type: EGCN
overlook_rates=[0.0, 0.0, 0.0]
BP_count=5, eta=100

============ Start testing ============

======= ACC: 1.0000, F1-Score: 1.0000 =======


======= ACC: 0.7860, F1-Score: 0.7847 =======


**======= ACC: 0.8130, F1-Score: 0.8105 =======**

**Comment by aWeirdName** [comment]
You can try the following settings:

**layers = [
    LayerParam(256, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN,
               learning_rate=10**-2, order=1, max_iter=100, lam=10**-3, batch_size=2708),
    LayerParam(128, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN,
               learning_rate=10**-4, order=2, max_iter=40, lam=10**-3, batch_size=100),
]**

change this in the run_classification.py script, and this is the result:


============ Settings ============
Totally 2 layers:
1-th layer: Neurons: 256, inner_activation: <function relu>, activations: <function leaky_relu> with {'negative_slope': 0.2}  with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 100, 'lam': 0.001, 'batch_size': 2708} type: EGCN
2-th layer: Neurons: 128, inner_activation: <function relu>, activations: <function leaky_relu> with {'negative_slope': 0.2}  with parameters: {'learning_rate': 0.0001, 'order': 2, 'max_iter': 40, 'lam': 0.001, 'batch_size': 100} type: EGCN
overlook_rates=[0.0, 0.0]
BP_count=5, eta=100

============ Start testing ============

======= ACC: 1.0000, F1-Score: 1.0000 =======


======= ACC: 0.7820, F1-Score: 0.7720 =======


======= ACC: 0.8120, F1-Score: 0.8082 =======

**Comment by aWeirdName** [comment]
You can try the following settings:

layers = [
LayerParam(256, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN,
learning_rate=10-2, order=1, max_iter=100, lam=10**-3, batch_size=2708),
LayerParam(128, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN,
learning_rate=10**-4, order=2, max_iter=40, lam=10**-3, batch_size=100),
]

change this in the run_classification.py script, and this is the result:

============ Settings ============
Totally 2 layers:
1-th layer: Neurons: 256, inner_activation: , activations: with {'negative_slope': 0.2} with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 100, 'lam': 0.001, 'batch_size': 2708} type: EGCN
2-th layer: Neurons: 128, inner_activation: , activations: with {'negative_slope': 0.2} with parameters: {'learning_rate': 0.0001, 'order': 2, 'max_iter': 40, 'lam': 0.001, 'batch_size': 100} type: EGCN
overlook_rates=[0.0, 0.0]
BP_count=5, eta=100

============ Start testing ============

======= ACC: 1.0000, F1-Score: 1.0000 =======

======= ACC: 0.7820, F1-Score: 0.7720 =======

======= ACC: 0.8120, F1-Score: 0.8082 =======

**Comment by liyhc** [comment]
> You can try the following settings:
> 
> layers = [ LayerParam(256, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN, learning_rate=10-2, order=1, max_iter=100, lam=10**-3, batch_size=2708), LayerParam(128, inner_act=relu_func, act=leaky_relu_func, gnn_type=LayerParam.EGCN, learning_rate=10**-4, order=2, max_iter=40, lam=10**-3, batch_size=100), ]
> 
> change this in the run_classification.py script, and this is the result:
> 
> ============ Settings ============ Totally 2 layers: 1-th layer: Neurons: 256, inner_activation: , activations: with {'negative_slope': 0.2} with parameters: {'learning_rate': 0.01, 'order': 1, 'max_iter': 100, 'lam': 0.001, 'batch_size': 2708} type: EGCN 2-th layer: Neurons: 128, inner_activation: , activations: with {'negative_slope': 0.2} with parameters: {'learning_rate': 0.0001, 'order': 2, 'max_iter': 40, 'lam': 0.001, 'batch_size': 100} type: EGCN overlook_rates=[0.0, 0.0] BP_count=5, eta=100
> 
> ============ Start testing ============
> 
> ======= ACC: 1.0000, F1-Score: 1.0000 =======
> 
> ======= ACC: 0.7820, F1-Score: 0.7720 =======
> 
> ======= ACC: 0.8120, F1-Score: 0.8082 =======

ok, I successfully replicated the results on the Cora dataset following your code
